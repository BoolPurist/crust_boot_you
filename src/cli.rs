pub use subcommands::SubCommands;

mod subcommands;

pub use from_cli_path::AbsoluteExistingPath;
pub use subcommands::SaveTemplateCli;
pub use subcommands::TemplateCliArg;

mod from_cli_path;

use clap::Parser;

#[derive(Debug, Parser)]
pub struct AppCliEntry {
    #[clap(subcommand)]
    sub_commands: SubCommands,
}

impl AppCliEntry {
    pub fn sub_commands(&self) -> &SubCommands {
        &self.sub_commands
    }
}
