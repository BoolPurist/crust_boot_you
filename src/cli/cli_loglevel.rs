use clap::ValueEnum;
use log::LevelFilter;

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, ValueEnum)]
pub enum CliLogLevel {
    None,
    Error,
    Warn,
    Info,
    Debug,
    Trace,
}

impl From<CliLogLevel> for LevelFilter {
    fn from(value: CliLogLevel) -> Self {
        match value {
            CliLogLevel::None => LevelFilter::Off,
            CliLogLevel::Error => LevelFilter::Error,
            CliLogLevel::Warn => LevelFilter::Warn,
            CliLogLevel::Info => LevelFilter::Info,
            CliLogLevel::Debug => LevelFilter::Debug,
            CliLogLevel::Trace => LevelFilter::Trace,
        }
    }
}
