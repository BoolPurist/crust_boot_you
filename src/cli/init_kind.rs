use std::{fmt::Display, str::FromStr};

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum InitKind {
    OnlyEmpty,
    NoNameConflicts,
    Override,
    Purge,
}

impl Default for InitKind {
    fn default() -> Self {
        Self::OnlyEmpty
    }
}
use crate::constants::{
    INIT_KIND_NO_NAME_CONFLICT, INIT_KIND_ONLY_EMPTY, INIT_KIND_OVERRIDE, INIT_KIND_PURGE,
};

pub const fn valid_values_init_kind_msg() -> &'static str {
    const_format::formatcp!(
        "Valid values are ({}, {}, {}, {})",
        INIT_KIND_ONLY_EMPTY,
        INIT_KIND_NO_NAME_CONFLICT,
        INIT_KIND_OVERRIDE,
        INIT_KIND_PURGE
    )
}

impl Display for InitKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let output = match self {
            Self::OnlyEmpty => INIT_KIND_ONLY_EMPTY,
            Self::NoNameConflicts => INIT_KIND_NO_NAME_CONFLICT,
            Self::Override => INIT_KIND_OVERRIDE,
            Self::Purge => INIT_KIND_PURGE,
        };
        write!(f, "{}", output)
    }
}
impl FromStr for InitKind {
    type Err = String;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            INIT_KIND_ONLY_EMPTY => Ok(Self::OnlyEmpty),
            INIT_KIND_NO_NAME_CONFLICT => Ok(Self::NoNameConflicts),
            INIT_KIND_OVERRIDE => Ok(Self::Override),
            INIT_KIND_PURGE => Ok(Self::Purge),
            _ => Err(format!(
                "{} is not one of the valid words. {}",
                s,
                valid_values_init_kind_msg()
            )),
        }
    }
}
