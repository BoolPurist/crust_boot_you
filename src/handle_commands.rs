use crate::{
    cli::SaveTemplateCli,
    file_management::{self, FileKind},
    prelude::*,
    AppCliEntry, NotEmptyText, SubCommands,
};
use std::path::Path;

pub type ReturnType = AppResult<String>;

pub fn handle(
    path_provider: impl PathProvider,
    file_manipulator: impl FileManipulator,
    args: &AppCliEntry,
) -> ReturnType {
    match args.sub_commands() {
        SubCommands::LoadTemplate { name } => {
            handle_load_template(path_provider, file_manipulator, name)
        }
        SubCommands::SaveTemplate(args) => {
            handle_save_template(path_provider, file_manipulator, args)
        }
    }
}

fn handle_load_template(
    path_provider: impl PathProvider,
    file_manipulator: impl FileManipulator,
    name: &NotEmptyText,
) -> ReturnType {
    debug!("Handling subcommand: {:?}", "LoadTemplate");

    let path_to_template = path_provider.specific_entry_template_files(name)?;

    if !path_to_template.exists() {
        bail!("No content for teamplate named {}, could be found", name);
    }

    let cwd = path_provider.cwd()?;
    file_manipulator.copy_dir(path_to_template, &cwd)?;

    Ok(format!(
        "Folder {:?} filled with content from Template ({})",
        cwd, name
    ))
}

fn save_template(
    path_provider: impl PathProvider,
    file_manipulator: impl FileManipulator,
    name: &NotEmptyText,
    path: impl AsRef<Path>,
) -> ReturnType {
    debug!("Handling subcommand: {:?}", "SaveTemplate");
    let (source_path, template_name) = (path.as_ref(), name);
    info!("Copying from {:?}", source_path);
    let file_type = file_management::detect_file_kind(source_path)?;

    match file_type {
        FileKind::File => {
            debug!("Detected {:?} as file", source_path);
            let ensured_template_folder = file_management::ensure_target_template_folder(
                &path_provider,
                &file_manipulator,
                template_name,
            )?;

            let target_path =
                file_management::construct_file_target_path(source_path, &ensured_template_folder)?;
            debug!(
                "Copying from source path {:?} to target path {:?}",
                source_path, target_path,
            );
            file_manipulator.copy_file(source_path, target_path)?;
        }
        FileKind::Folder => {
            debug!("Detected {:?} as folder", source_path);
            let target_path = file_management::ensure_target_template_folder(
                &path_provider,
                &file_manipulator,
                template_name,
            )?;
            debug!(
                "Copying from source path {:?} to target path {:?}",
                source_path, target_path,
            );
            file_manipulator.copy_dir(source_path, target_path)?;
        }
    }

    let msg_to_user = format!(
        "Created template with name ({}) from the {:?}",
        name, source_path,
    );
    Ok(msg_to_user)
}
fn handle_save_template(
    path_provider: impl PathProvider,
    file_manipulator: impl FileManipulator,
    from_cli: &SaveTemplateCli,
) -> ReturnType {
    let arguments = from_cli.arguments();
    let (name, path) = (arguments.name(), arguments.path());
    save_template(path_provider, file_manipulator, name, path)
}
