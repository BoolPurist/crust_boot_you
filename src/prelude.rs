use std::io;

pub use crate::constants;
pub use anyhow::Context;

pub type IoResult<T = ()> = Result<T, io::Error>;
pub type AppError = anyhow::Error;
pub type AppResult<T = ()> = Result<T, AppError>;

pub type ReturnToUser = AppResult<String>;
pub type PathResult = AppResult<PathBuf>;
pub type AppIoResult<T = ()> = Result<T, AppIoError>;

pub use crate::app_traits::file_manipulator::FileManipulator;
pub use crate::app_traits::path_provider::PathProvider;
pub use crate::file_management::AppIoError;
pub use crate::NotEmptyText;
pub use std::path::{Path, PathBuf};

#[cfg(test)]
pub use mockall::automock;
