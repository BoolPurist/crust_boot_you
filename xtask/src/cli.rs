use clap::Parser;
mod subcommands;
pub use subcommands::TaskSubcommand;

#[derive(Debug, Parser)]
pub struct TaskCliEntry {
    #[clap(subcommand)]
    subcommands: TaskSubcommand,
}

impl TaskCliEntry {
    pub fn subcommands(&self) -> &TaskSubcommand {
        &self.subcommands
    }
}
