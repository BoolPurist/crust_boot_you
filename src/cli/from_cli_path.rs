use crate::prelude::*;
use derive_more::Into;
use std::{
    path::{Path, PathBuf},
    str::FromStr,
};

#[derive(Debug, Clone)]
pub struct AbsoluteExistingDirPath(PathBuf);

impl FromStr for AbsoluteExistingDirPath {
    type Err = AppError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let existing: AbsoluteExistingPath = s.parse()?;
        Self::new(existing)
    }
}

impl AbsoluteExistingDirPath {
    pub fn new(absolute_existing: AbsoluteExistingPath) -> AppResult<Self> {
        let absolute_existing: PathBuf = absolute_existing.into();

        ensure!(
            absolute_existing.is_dir(),
            anyhow!(
                "Path does not point to a directory at {:?}",
                &absolute_existing
            )
        );
        Ok(Self(absolute_existing))
    }
    pub fn as_path(&self) -> &Path {
        self.0.as_path()
    }
}

#[derive(Debug, Clone, Into)]
pub struct AbsoluteExistingPath(PathBuf);

impl AbsoluteExistingPath {
    pub fn new(path: PathBuf, resolver: &impl PathResolver) -> AppResult<Self> {
        let new_p = resolver
            .absolute_and_exits(&path)?
            .ok_or_else(|| anyhow!("Path {:?} does not exit !.\n", path))?
            .to_path_buf();
        Ok(Self(new_p))
    }
}

impl FromStr for AbsoluteExistingPath {
    type Err = AppError;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let path = s.into();
        Self::new(path, &*constants::USED_PATH_PROVIDER)
    }
}

impl std::ops::Deref for AbsoluteExistingPath {
    type Target = Path;

    fn deref(&self) -> &Self::Target {
        self.as_ref()
    }
}

impl AsRef<Path> for AbsoluteExistingPath {
    fn as_ref(&self) -> &Path {
        self.0.as_path()
    }
}
