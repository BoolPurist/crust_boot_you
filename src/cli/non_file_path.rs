use std::{
    path::{Path, PathBuf},
    str::FromStr,
};

use crate::prelude::{AppError, AppResult};

#[derive(Debug, Clone)]
pub struct NonFilePath(PathBuf);

impl FromStr for NonFilePath {
    type Err = AppError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let path = PathBuf::from(s);
        Self::new(path)
    }
}

impl NonFilePath {
    pub fn new(path: PathBuf) -> AppResult<Self> {
        let exits = path.try_exists()?;
        ensure!(
            exits && !path.is_file(),
            anyhow!("Path must not point to {:?} a file", &path)
        );
        Ok(Self(path))
    }
    pub fn as_path(&self) -> &Path {
        self.0.as_path()
    }
}
