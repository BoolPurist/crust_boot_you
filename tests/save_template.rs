mod common;
use common::insta_utils;
use common::prelude::*;
use crust_boot_you::cli::{AbsoluteExistingPath, SaveTemplateCli};
use crust_boot_you::handle_commands;

#[test]
#[named]
fn save_file_new_template() {
    let setup = TestSetup::new(actual_expected!());

    let given_name = ValidTemplateName::new("a".to_string()).unwrap();
    let given_path =
        AbsoluteExistingPath::new(PathBuf::from("new.txt"), setup.path_resolver()).unwrap();
    let arguments = SaveTemplateCli::new(given_name, given_path);
    let output =
        handle_commands::handle_save_template(setup.path_provider(), setup.os_mani(), &arguments)
            .expect("Should be successful in this test case");
    insta::with_settings!({ filters => insta_utils::filter_random_tmp_folder_name() }, { insta::assert_display_snapshot!(output) });

    setup.assert_with_expected();
}

#[test]
#[named]
fn save_folder_as_template() {
    let setup = TestSetup::new(actual_expected!());

    let given_name = ValidTemplateName::new("some hello project".to_string()).unwrap();
    let given_path =
        AbsoluteExistingPath::new(PathBuf::from("lang").join("go"), setup.path_resolver()).unwrap();
    let arguments = SaveTemplateCli::new(given_name, given_path);
    let output =
        handle_commands::handle_save_template(setup.path_provider(), setup.os_mani(), &arguments)
            .expect("Should be successful in this test case");

    setup.assert_with_expected();

    insta::with_settings!({ filters => insta_utils::filter_random_tmp_folder_name() }, { insta::assert_display_snapshot!(output) });
}

#[test]
#[named]
fn save_template_err_source_path_no_existence() {
    let setup = TestSetup::new(actual_expected!());

    let _ = AbsoluteExistingPath::new(PathBuf::from("lang").join("go"), setup.path_resolver())
        .expect_err("Should be an error in this test case");

    setup.assert_with_expected();
}

#[test]
#[named]
fn save_template_err_template_already_there() {
    let setup = TestSetup::new(actual_expected!());

    let given_name = ValidTemplateName::new("already_there".to_string()).unwrap();
    let given_path =
        AbsoluteExistingPath::new(PathBuf::from("project"), setup.path_resolver()).unwrap();
    let arguments = SaveTemplateCli::new(given_name, given_path);
    let output =
        handle_commands::handle_save_template(setup.path_provider(), setup.os_mani(), &arguments)
            .expect_err("Should be an error in this test case");
    setup.assert_with_expected();
    insta::assert_display_snapshot!(output);
}
