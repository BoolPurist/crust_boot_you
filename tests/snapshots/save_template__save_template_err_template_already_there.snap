---
source: tests/save_template.rs
expression: output
---
Template with the name (already_there) is already created
